File: 2080-range-frequency-queries/2080-range-frequency-queries.py
class RangeFreqQuery:

    def __init__(self, arr: List[int]):
        self.arr = arr
        self.hash = dict()
        for i, v in enumerate(arr):
            if v not in self.hash:
                self.hash[v] = []
            self.hash[v].append(i)

    def query(self, left: int, right: int, value: int) -> int:
        if value not in self.hash:
            return 0
        l = bisect.bisect_left(self.hash[value], left)
        r = bisect.bisect_right(self.hash[value], right)

        return r - l
        
        


# Your RangeFreqQuery object will be instantiated and called as such:
# obj = RangeFreqQuery(arr)
# param_1 = obj.query(left,right,value)

File: 2080-range-frequency-queries/2080-range-frequency-queries.java
class RangeFreqQuery {
    private final Map<Integer, ArrayList<Integer>> hash = new HashMap<>();
    public RangeFreqQuery(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            if (!hash.containsKey(arr[i])) {
                hash.put(arr[i], new ArrayList<>());
            }
            hash.get(arr[i]).add(i);
        }
    }
    
    public int query(int left, int right, int value) {
        if (!hash.containsKey(value)) {
            return 0;
        }

        int l = lowerBound(left, hash.get(value));
        int r = upperBound(right, hash.get(value));

        // System.out.println("" + l + ", " + r);

        return r - l;
    }

    private int lowerBound(int x, ArrayList<Integer> arr) {
        int l = 0, r = arr.size();
        while (l < r) {
            int mid = (l + r) / 2;
            if (arr.get(mid) < x) {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        return l;
    }

    private int upperBound(int x, ArrayList<Integer> arr) {
        int l = 0, r = arr.size();
        while (l < r) {
            int mid = (l + r) / 2;
            if (arr.get(mid) <= x) {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        return l;
    }
}

/**
 * Your RangeFreqQuery object will be instantiated and called as such:
 * RangeFreqQuery obj = new RangeFreqQuery(arr);
 * int param_1 = obj.query(left,right,value);
 */